src/bin/config-store.rs

```diff
@@ -0,0 +1,116 @@
+    command: Commands,
+}
+
+#[derive(Debug, Subcommand)]
+enum Commands {
+    #[clap(arg_required_else_help = true, about = "View a setting")]
+    View {
+        #[clap(required = true, short = 'k', long = "key")]
+        key: String,
+    },
+    #[clap(about = "List all settings")]
+    List,
+    #[clap(arg_required_else_help = true, about = "Set a setting")]
+    Set {
+        #[clap(required = true, short = 'k', long = "key")]
+        key: String,
+        #[clap(required = true, short = 'v', long = "value")]
+        value: String,
+    },
+    #[clap(arg_required_else_help = true, about = "Search for a setting")]
+    Search {
+        #[clap(required = true, short = 'k', long = "key")]
+        key: String,
+    },
+}
+
+#[derive(Debug, Clone, Copy, clap::ValueEnum, Display)]
+enum Engine {
+    Sqlite,
+    Json,
+}
+
+impl std::str::FromStr for Engine {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "sqlite" => Ok(Engine::Sqlite),
+            "json" => Ok(Engine::Json),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Debug, Parser)]
+struct GlobalOpts {
+    #[clap(short = 'e', long = "engine", global = true, default_value = "sqlite")]
+    engine: Engine,
+    #[clap(
+        short = 'p',
+        long = "path",
+        global = true,
+        default_value = "settings.db"
+    )]
+    path: String,
+}
+
+fn main() {
+    let args = Cli::parse();
+
+    let storage_box: Box<dyn StorageAdapter> = match args.global_opts.engine {
+        Engine::Sqlite => Box::new(SqliteStorageAdapter::new(args.global_opts.path.as_str())),
+        Engine::Json => Box::new(JsonStorageAdapter::new(args.global_opts.path.as_str())),
+    };
+
+    let mut store = ConfigStore::new(storage_box, true);
+
+    match args.command {
+        Commands::View { key } => {
+            if !store.has(key.as_str()) {
+                println!("Key not found");
+                return;
+            }
+
+            let info = store.get_info(key.as_str()).unwrap();
+            let value = store.get(key.as_str());
+
+            println!("Key            : {}", key);
+            println!("Current Value  : {}", value);
+            println!("Default Value  : {}", info.default);
+            println!("Description    : {}", info.description);
+        }
+        Commands::List => {
+            for key in store.find("*") {
+                let value = store.get(key.as_str());
+                println!("{:40}: {}", key, value);
+            }
+        }
+        Commands::Set { key, value } => {
+            store.set(
+                key.as_str(),
+                Setting::from_string(value.as_str()).expect("incorrect value"),
+            );
+        }
+        Commands::Search { key } => {
+            for key in store.find(key.as_str()) {
+                let value = store.get(key.as_str());
+                println!("{:40}: {}", key, value);
+            }
+        }
+    }
+}
```

Comment:
I would add this line to the `Makefile` for this command:

```patch
diff --git a/Makefile b/Makefile
index e25ba44..7a362d3 100644
--- a/Makefile
+++ b/Makefile
@@ -36,6 +36,7 @@ test_fmt:
 test_commands:
        cargo run --bin html5-parser-test >/dev/null
        cargo run --bin parser-test >/dev/null
+       cargo run --bin config-store list >/dev/null
 
 help: ## Display available commands
        echo "Available make commands:"
```

Reply:
I would also add `settings.db` to `.gitignore` or put it under a `tmp/` directory that is ignored:

```patch
diff --git a/.gitignore b/.gitignore
index 5ce6f87..0a3c97f 100644
--- a/.gitignore
+++ b/.gitignore
@@ -5,4 +5,5 @@ target/
 *.swp*
 
 # Local testing 
-local/
\ No newline at end of file
+local/
+settings.db
```

![2023-11-11_11-05](https://github.com/gosub-browser/gosub-engine/assets/760949/0e6ff592-43ae-473a-b711-fe236b1d398c)